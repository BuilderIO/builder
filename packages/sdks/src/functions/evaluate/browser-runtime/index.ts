export { runInBrowser as evaluator } from './browser.js';
