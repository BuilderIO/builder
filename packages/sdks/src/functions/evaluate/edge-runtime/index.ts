export { runInEdge as evaluator } from './edge-runtime.js';
