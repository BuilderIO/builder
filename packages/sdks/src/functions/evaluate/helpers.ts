import type {
  BuilderContextInterface,
  BuilderRenderState,
} from '../../context/types.js';
import type { getUserAttributes } from '../track/helpers.js';

export type BuilderGlobals = {
  isEditing: boolean | undefined;
  isBrowser: boolean | undefined;
  isServer: boolean | undefined;
  getUserAttributes: typeof getUserAttributes;
};

export type ExecutorArgs = Pick<
  BuilderContextInterface,
  'localState' | 'context' | 'rootState' | 'rootSetState'
> & {
  code: string;
  builder: BuilderGlobals;
  event: Event | undefined;
};

export type Executor = (args: ExecutorArgs) => any;

export type FunctionArguments = ReturnType<typeof getFunctionArguments>;

export const getFunctionArguments = ({
  builder,
  context,
  event,
  state,
}: Pick<ExecutorArgs, 'builder' | 'context' | 'event'> & {
  state: BuilderRenderState;
}) => {
  return Object.entries({
    state,
    Builder: builder, // legacy
    builder,
    context,
    event,
  });
};
