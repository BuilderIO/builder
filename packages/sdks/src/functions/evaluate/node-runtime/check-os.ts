import fs from 'fs';
import os from 'os';

let cachedOS: string | null = null;

const whichOS = (): string => {
  if (cachedOS) return cachedOS;

  const platform = os.platform();
  switch (platform) {
    case 'linux':
      try {
        const data = fs.readFileSync('/etc/os-release', 'utf8');
        const nameMatch = data.match(/^NAME="(.+)"$/m);
        const versionMatch = data.match(/^VERSION_ID="(.+)"$/m);
        const osName = nameMatch?.[1] ?? 'Linux (unknown distribution)';
        const osVersion = versionMatch?.[1] ?? 'unknown version';
        cachedOS = `${osName} ${osVersion}`;
      } catch (err) {
        console.warn(
          `Error reading Linux OS information from /etc/os-release:`,
          err
        );
        cachedOS = 'Linux';
      }
      break;
    case 'darwin':
      cachedOS = 'macOS';
      break;
    case 'win32':
      cachedOS = 'Windows';
      break;
    default:
      cachedOS = platform;
  }

  return cachedOS;
};

const osName = whichOS();
console.log('[Builder.io] running on OS:', osName);

const ubuntuVersionMatch = osName.match(/Ubuntu (\d+)/);
if (ubuntuVersionMatch) {
  const ubuntuVersion = parseInt(ubuntuVersionMatch[1]);
  if (ubuntuVersion >= 22) {
    console.warn(
      `WARNING: You are running Builder SDK on ${osName}. This is known to cause crashes with \`isolated-vm\`. See https://github.com/BuilderIO/builder/issues/3605 for more information.`
    );
  }
}
