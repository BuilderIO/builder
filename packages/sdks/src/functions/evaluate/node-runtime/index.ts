export { evaluator } from '../browser-runtime/index.js';
