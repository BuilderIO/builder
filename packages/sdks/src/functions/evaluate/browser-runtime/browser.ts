import type { BuilderRenderState } from '../../../context/types.js';
import type { ExecutorArgs } from '../helpers.js';
import { getFunctionArguments } from '../helpers.js';

export const runInBrowser = ({
  code,
  builder,
  context,
  event,
  localState,
  rootSetState,
  rootState,
}: ExecutorArgs) => {
  const functionArgs = getFunctionArguments({
    builder,
    context,
    event,
    state: flattenState({ rootState, localState, rootSetState }),
  });

  return new Function(...functionArgs.map(([name]) => name), code)(
    ...functionArgs.map(([, value]) => value)
  );
};

export function flattenState({
  rootState,
  localState,
  rootSetState,
}: {
  rootState: Record<string | symbol, any>;
  localState: Record<string | symbol, any> | undefined;
  rootSetState: ((rootState: BuilderRenderState) => void) | undefined;
}): BuilderRenderState {
  return new Proxy(rootState, {
    get: (target, prop) => {
      if (localState && prop in localState) {
        return localState[prop];
      }

      const val = target[prop];
      if (typeof val === 'object' && val !== null) {
        return flattenState({
          rootState: val,
          localState: undefined,
          rootSetState: rootSetState
            ? (subState) => {
                target[prop] = subState;
                rootSetState(target);
              }
            : undefined,
        });
      }

      return val;
    },
    set: (target, prop, value) => {
      if (localState && prop in localState) {
        throw new Error(
          'Writing to local state is not allowed as it is read-only.'
        );
      }

      target[prop] = value;

      rootSetState?.(target);
      return true;
    },
  });
}
