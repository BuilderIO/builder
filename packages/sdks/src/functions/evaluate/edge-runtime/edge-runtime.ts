import { logger } from '../../../helpers/logger.js';
import type { ExecutorArgs } from '../helpers.js';
import { flattenState, getFunctionArguments } from '../helpers.js';
import Interpreter from './acorn-interpreter.js';

/**
 * https://stackoverflow.com/a/46503625
 */
function patchInterpreter() {
  const originalGetProperty = Interpreter.prototype.getProperty;
  const originalSetProperty = Interpreter.prototype.setProperty;

  function newGetProperty(this: typeof Interpreter, obj: any, name: any) {
    if (obj == null || !obj._connected) {
      return originalGetProperty.call(this, obj, name);
    }

    const value = obj._connected[name];
    if (Array.isArray(value)) {
      return this.nativeToPseudo(value);
    }

    if (typeof value === 'object') {
      // if the value is an object itself, create another connected object
      return this.createConnectedObject(value);
    }
    return value;
  }
  function newSetProperty(
    this: typeof Interpreter,
    obj: any,
    name: any,
    value: any,
    opt_descriptor: any
  ) {
    if (obj == null || !obj._connected) {
      return originalSetProperty.call(this, obj, name, value, opt_descriptor);
    }

    obj._connected[name] = this.pseudoToNative(value);
  }

  const getKeys: string[] = [];
  const setKeys: string[] = [];
  for (const key of Object.keys(Interpreter.prototype)) {
    if (Interpreter.prototype[key] === originalGetProperty) {
      getKeys.push(key);
    }
    if (Interpreter.prototype[key] === originalSetProperty) {
      setKeys.push(key);
    }
  }

  for (const key of getKeys) {
    Interpreter.prototype[key] = newGetProperty;
  }
  for (const key of setKeys) {
    Interpreter.prototype[key] = newSetProperty;
  }

  Interpreter.prototype.createConnectedObject = function (obj: any) {
    const connectedObject = this.createObject(this.OBJECT);
    connectedObject._connected = obj;
    return connectedObject;
  };
}
patchInterpreter();

const processCode = (code: string) => {
  return (
    code
      // strip async/await polyfill: remove anything before 'function main()'
      .replace(
        /^.*?function main\(\)/,
        `
var __awaiter = function (e, t, n, r) {return r()},
__generator = function (e, t) { return t() };
function main()`
      )
      // replace ?. with .
      .replace(/\?\./g, '.')
  );
};

const getJSONValName = (val: string) => val + 'JSON';
export const runInEdge = ({
  builder,
  context,
  event,
  rootState,
  localState,
  rootSetState,
  code,
}: ExecutorArgs) => {
  const state = flattenState({
    rootState,
    localState,
    rootSetState,
  });

  const properties = getFunctionArguments({
    builder,
    context,
    event,
    state,
  });

  /**
   * Deserialize all properties from JSON strings to JS objects
   */
  const prependedCode = properties
    .map(([key]) => {
      const jsonValName = getJSONValName(key);
      if (key === 'state') {
        return ``;
      }
      return `var ${key} = ${jsonValName} === undefined ? undefined : JSON.parse(${jsonValName});`;
    })
    .join('\n');
  const cleanedCode = processCode(code);
  if (cleanedCode === '') {
    logger.warn('Skipping evaluation of empty code block.');
    return;
  }

  const transformed = `
function theFunction() {
  ${prependedCode}

  ${cleanedCode}
}
theFunction();
`;

  const initFunc = function (interpreter: any, globalObject: any) {
    /**
     * serialize all function args to JSON strings
     */
    properties.forEach(([key, val]) => {
      if (key === 'state') {
        interpreter.setProperty(
          globalObject,
          key,
          interpreter.createConnectedObject(val),
          interpreter.READONLY_DESCRIPTOR
        );
      } else {
        const jsonVal = JSON.stringify(val);
        interpreter.setProperty(globalObject, getJSONValName(key), jsonVal);
      }
    });
  };

  const myInterpreter = new Interpreter(transformed, initFunc);
  myInterpreter.run();
  const output = myInterpreter.pseudoToNative(myInterpreter.value);

  return output;
};
