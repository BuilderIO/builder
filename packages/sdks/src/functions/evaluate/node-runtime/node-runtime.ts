import { fastClone } from '../../fast-clone.js';
import { set } from '../../set.js';
import type {
  BuilderGlobals,
  ExecutorArgs,
  FunctionArguments,
} from '../helpers.js';
import { getFunctionArguments } from '../helpers.js';
import { safeDynamicRequire } from './safeDynamicRequire.js';
const ivm: typeof import('isolated-vm') = safeDynamicRequire('isolated-vm');

const getSyncValName = (key: string) => `bldr_${key}_sync`;

const BUILDER_SET_STATE_NAME = 'BUILDER_SET_STATE';

const INJECTED_IVM_GLOBAL = 'BUILDER_IVM';

// Convert all argument references to proxies, and pass `copySync` method to target object, to return a copy of the original JS object
// https://github.com/laverdet/isolated-vm#referencecopysync
const REF_TO_PROXY_FN = `
var refToProxy = (obj) => {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }
  return new Proxy({}, {
    get(target, key) {
        if (key === 'copySync') {
          return () => obj.copySync();
        }
        const val = obj.getSync(key);
        if (typeof val?.getSync === 'function') {
            return refToProxy(val);
        }
        return val;
    },
    set(target, key, value) {
        const v = typeof value === 'object' ? new ${INJECTED_IVM_GLOBAL}.Reference(value) : value;
        obj.setSync(key, v);
        ${BUILDER_SET_STATE_NAME}(key, value)
    },
    deleteProperty(target, key) {
        obj.deleteSync(key);
    }
  })
}
`;
const processCode = ({
  code,
  args,
}: {
  code: string;
  args: FunctionArguments;
}) => {
  const fnArgs = args
    .map(([name]) => `var ${name} = refToProxy(${getSyncValName(name)}); `)
    .join('');

  // the output is stringified and parsed back to the parent isolate if needed (when it's an `object`)
  return `
${REF_TO_PROXY_FN}
${fnArgs}
function theFunction() {
  ${code}
}

let output = theFunction()

if (typeof output === 'object' && output !== null) {
  output = JSON.stringify(output.copySync ? output.copySync() : output);
}

output;
`;
};
const getIsolateContext = () => {
  // if (Builder.serverContext) {
  //   return Builder.serverContext;
  // }
  // Builder.setServerContext(isolate.createContextSync());
  const isolate = new ivm.Isolate({
    memoryLimit: 128,
  });
  return isolate.createContextSync();
};
export const runInNode = ({
  code,
  builder,
  context,
  event,
  localState,
  rootSetState,
  rootState,
}: ExecutorArgs) => {
  const state = fastClone({
    ...rootState,
    ...localState,
  });
  const args = getFunctionArguments({
    builder,
    context,
    event,
    state,
  });
  const isolateContext = getIsolateContext();
  const jail = isolateContext.global;
  // This makes the global object available in the context as `global`. We use `derefInto()` here
  // because otherwise `global` would actually be a Reference{} object in the new isolate.
  jail.setSync('global', jail.derefInto());

  // We will create a basic `log` function for the new isolate to use.
  jail.setSync('log', function (...logArgs: any[]) {
    console.log(...logArgs);
  });

  /**
   * Propagate state changes back to the reactive root state.
   */
  jail.setSync(BUILDER_SET_STATE_NAME, function (key: string, value: any) {
    // mutate the `rootState` object itself. Important for cases where we do not have `rootSetState`
    // like Qwik.
    set(rootState, key, value);
    // call the `rootSetState` function if it exists
    rootSetState?.(rootState);
  });
  args.forEach(([key, arg]) => {
    const val =
      typeof arg === 'object'
        ? new ivm.Reference(
            // workaround: methods with default values for arguments is not being cloned over
            key === 'builder'
              ? {
                  ...arg,
                  getUserAttributes: () =>
                    (arg as BuilderGlobals).getUserAttributes(),
                }
              : arg
          )
        : null;
    jail.setSync(getSyncValName(key), val);
  });

  jail.setSync(INJECTED_IVM_GLOBAL, ivm);

  const evalStr = processCode({
    code,
    args,
  });
  const resultStr = isolateContext.evalSync(evalStr);
  try {
    // returning objects throw errors in isolated vm, so we stringify it and parse it back
    const res = JSON.parse(resultStr);
    return res;
  } catch (_error: any) {
    return resultStr;
  }
};
