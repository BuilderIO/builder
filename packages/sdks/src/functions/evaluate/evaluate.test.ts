import { runInBrowser } from './browser-runtime/browser';
import { runInEdge } from './edge-runtime/edge-runtime';
import type { ExecutorArgs } from './helpers';
import { getBuilderGlobals, parseCode } from './helpers';
import { runInNode } from './node-runtime/node-runtime';

const DEFAULTS = {
  builder: getBuilderGlobals(),
  context: {},
  event: undefined,
  localState: {},
  rootSetState: () => {},
  rootState: {},
};

const getEval = () => {
  switch (process.env.SDK_ENV) {
    case 'browser':
      return runInBrowser;
    case 'node':
      return runInNode;
    case 'edge':
      return runInEdge;
    default:
      throw new Error('Invalid SDK_ENV');
  }
};
const evaluateCode = (args: ExecutorArgs) => {
  try {
    return getEval()(args);
  } catch (e: any) {
    console.error('Failed code evaluation: ' + e.message, args.code);
    throw e;
  }
};

const TESTS = {
  'does simple math': () => {
    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode('1 + 1', { isExpression: true }),
    });

    expect(output).toBe(2);
  },
  'returns state value': () => {
    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode('state.foo', { isExpression: true }),
      rootState: {
        foo: 'bar',
      },
    });

    expect(output).toBe('bar');
  },
  'returns nested state value': () => {
    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode('state.foo.bar.baz', { isExpression: true }),
      rootState: {
        foo: {
          bar: {
            baz: 'qux',
          },
        },
      },
    });

    expect(output).toBe('qux');
  },
  'returns nested local state value': () => {
    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode('state.foo[0].bar.baz', { isExpression: true }),
      localState: {
        foo: [
          {
            bar: {
              baz: 'qux',
            },
          },
        ],
      },
    });

    expect(output).toBe('qux');
  },
  'sets simple state value (using `rootSetState`)': () => {
    const rootState = {
      foo: 'bar',
    };

    runInNode({
      ...DEFAULTS,
      code: parseCode('state.foo = "baz"', { isExpression: true }),
      rootState,
      rootSetState: (newState) => {
        Object.assign(rootState, newState);
      },
    });

    expect(rootState.foo).toBe('baz');
  },
  'sets simple state value (if `rootState` is mutable)': () => {
    const rootState = {
      foo: 'bar',
    };

    runInNode({
      ...DEFAULTS,
      code: parseCode('state.foo = "baz"', { isExpression: true }),
      rootState,
      rootSetState: undefined,
    });

    expect(rootState.foo).toBe('baz');
  },
  'set & read simple state value (if `rootState` is mutable)': () => {
    const rootState: {
      a?: number;
      b?: number;
    } = {};

    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode(
        `
        state.a = 10;
        state.b = state.a + 4;
        var x = state.b;

        return x
      `,
        { isExpression: true }
      ),
      rootState,
      rootSetState: undefined,
    });

    expect(rootState.a).toBe(10);
    expect(rootState.b).toBe(14);
    expect(output).toBe(14);
  },
  'set & read simple state value (with `rootSetState`)': () => {
    const rootState: {
      a?: number;
      b?: number;
    } = {};

    const output = evaluateCode({
      ...DEFAULTS,
      code: parseCode(
        `
        state.a = 10;
        state.b = state.a + 4;
        var x = state.b;

        return x
      `,
        { isExpression: true }
      ),
      rootState,
      rootSetState: (newState) => {
        Object.assign(rootState, newState);
      },
    });

    expect(rootState.a).toBe(10);
    expect(rootState.b).toBe(14);
    expect(output).toBe(14);
  },
  'set Array state value (if `rootState` is mutable)': () => {
    const rootState: { a?: number[] } = {};

    runInNode({
      ...DEFAULTS,
      code: parseCode(`state.a = [10, 43];`, { isExpression: true }),
      rootState,
    });

    expect(rootState.a).toBeInstanceOf(Array);
    expect(rootState.a).toHaveLength(2);
    expect(rootState.a?.[0]).toBe(10);
    expect(rootState.a?.[1]).toBe(43);
  },
};

describe(`evaluate (${process.env.SDK_ENV})`, () => {
  describe('individual evaluations', () => {
    Object.keys(TESTS).forEach((key) => {
      test(key, () => {
        TESTS[key as keyof typeof TESTS]();
      });
    });
  });
  test('successfully runs multiple evaluations in random order', () => {
    const keys = Object.keys(TESTS);
    const randomKeys = keys.sort(() => Math.random() - 0.5);

    randomKeys.forEach((key) => {
      TESTS[key as keyof typeof TESTS]();
    });
  });
});
