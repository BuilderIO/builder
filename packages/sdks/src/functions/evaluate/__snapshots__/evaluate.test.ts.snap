// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSimpleExpressionGetPath > returns the path from "return state.foo.bar" 1`] = `"foo.bar"`;

exports[`getSimpleExpressionGetPath > returns the path from "return state.foo.bar;" 1`] = `"foo.bar"`;

exports[`getSimpleExpressionGetPath > returns the path from "state.foo.bar" 1`] = `"foo.bar"`;

exports[`getSimpleExpressionGetPath > returns the path from "state.foo.bar;" 1`] = `"foo.bar"`;

exports[`getSimpleExpressionGetPath > returns the path from "var _virtual_index=state.foo.bar ;return _virtual_index" 1`] = `"foo.bar"`;

exports[`getSimpleExpressionGetPath > returns the path from "var _virtual_index=state.foo.bar;return _virtual_index" 1`] = `"foo.bar"`;
